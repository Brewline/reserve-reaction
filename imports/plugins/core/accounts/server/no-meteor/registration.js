import addRolesToGroups from "/imports/plugins/core/core/server/Reaction/addRolesToGroups";

/**
 * @summary Will be called for every plugin
 * @param {Object} options The options object that the plugin passed to registerPackage
 * @returns {undefined}
 */
export function registerPluginHandler({ addRolesToGroups: packageRolesAndGroups }) {
  if (Array.isArray(packageRolesAndGroups)) {
    packageRolesAndGroups.forEach(({ groups, roles }) => {
      addRolesToGroups({ allShops: true, groups, roles });
    });
  }
}
