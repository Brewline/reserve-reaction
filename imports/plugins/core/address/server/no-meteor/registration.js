import { get } from "lodash";

export const addressValidationServices = {};

/**
 * @summary Will be called for every plugin
 * @param {Object} options The options object that the plugin passed to registerPackage
 * @returns {undefined}
 */
export function registerPluginHandler({
  name: pluginName,
  addressValidationServices: pluginAddressValidationServices
}) {
  if (Array.isArray(pluginAddressValidationServices)) {
    for (const pluginAddressValidationService of pluginAddressValidationServices) {
      addressValidationServices[pluginAddressValidationService.name] = {
        ...pluginAddressValidationService,
        pluginName
      };
    }
  }
}

/**
 * @param {Object} context The app context
 * @param {String} shopId The shop ID
 * @returns {Object[]} The enabled validation services for the shop with ID `shopId`
 */
export async function getEnabledAddressValidationServicesForShop(context, shopId) {
  const plugin = await context.collections.Packages.findOne({ name: "reaction-address", shopId });
  if (!plugin) return [];

  return get(plugin, "settings.addressValidation.enabledServices", []);
}

/**
 * @param {Object} context The app context
 * @param {String} shopId The shop ID
 * @param {String} countryCode The country code of the address to be validated
 * @returns {Object[]} The first enabled validation services for the shop with ID `shopId`
 *   that supports countryCode.
 */
export async function getAddressValidationService(context, shopId, countryCode) {
  const services = await getEnabledAddressValidationServicesForShop(context, shopId);
  const supportedServices = services.filter((ser) => {
    const registeredService = addressValidationServices[ser.serviceName];
    return (
      !Array.isArray(registeredService.supportedCountryCodes) ||
      registeredService.supportedCountryCodes.includes(countryCode)
    );
  });
  const foundService = supportedServices.find((service) => !Array.isArray(service.countryCodes) || service.countryCodes.includes(countryCode));
  if (!foundService) return null;
  return addressValidationServices[foundService.serviceName];
}
