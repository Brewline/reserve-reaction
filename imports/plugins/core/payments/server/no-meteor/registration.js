export const paymentMethods = {};

/**
 * @summary Will be called for every plugin
 * @param {Object} options The options object that the plugin passed to registerPackage
 * @returns {undefined}
 */
export function registerPluginHandler({ name: pluginName, paymentMethods: pluginPaymentMethods }) {
  if (Array.isArray(pluginPaymentMethods)) {
    for (const pluginPaymentMethod of pluginPaymentMethods) {
      paymentMethods[pluginPaymentMethod.name] = { ...pluginPaymentMethod, pluginName };
    }
  }
}

/**
 * @name getPaymentMethodConfigByName
 * @param {String} name payment method name, e.g. example, stripe_card
 * @returns {Object} payment method configuration
 */
export function getPaymentMethodConfigByName(name) {
  const config = paymentMethods[name];
  if (!config) {
    throw new Error(`Configuration not found for ${name} payment method. Did you remove the plugin that provides this payment method?`);
  }
  return config;
}
