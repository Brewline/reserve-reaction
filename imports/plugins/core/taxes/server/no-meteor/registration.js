export const taxServices = {};

/**
 * @summary Will be called for every plugin
 * @param {Object} options The options object that the plugin passed to registerPackage
 * @returns {undefined}
 */
export function registerPluginHandler({ name: pluginName, taxServices: pluginTaxServices }) {
  if (Array.isArray(pluginTaxServices)) {
    for (const pluginTaxService of pluginTaxServices) {
      taxServices[pluginTaxService.name] = { ...pluginTaxService, pluginName };
    }
  }
}

/**
 * @param {Object} context The app context
 * @param {String} shopId The shop ID
 * @returns {Object|null} The definition from registerPackage for the tax service that is
 *   currently enabled for the shop with ID `shopId`
 */
export async function getActiveTaxServiceForShop(context, shopId) {
  const plugin = await context.collections.Packages.findOne({ name: "reaction-taxes", shopId });
  if (!plugin) return null;

  const { activeTaxServiceName } = plugin.settings || {};
  if (!activeTaxServiceName) return null;

  const config = taxServices[activeTaxServiceName];
  if (!config) {
    throw new Error(`Active tax service is "${activeTaxServiceName}" but no such service exists. ` +
      "Did you forget to install the plugin that provides this service?");
  }

  return config;
}
