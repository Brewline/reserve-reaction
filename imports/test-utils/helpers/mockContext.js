const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
