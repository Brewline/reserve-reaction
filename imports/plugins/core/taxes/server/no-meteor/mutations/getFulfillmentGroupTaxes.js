import Logger from "@reactioncommerce/logger";
import ReactionError from "@reactioncommerce/reaction-error";
import { CommonOrder } from "/imports/plugins/core/orders/server/no-meteor/simpleSchemas";
import { getActiveTaxServiceForShop } from "../registration";
import { TaxServiceResult } from "../../../lib/simpleSchemas";

/**
 * @summary Returns all taxes that apply to a provided order, delegating to a more specific
 *   tax calculation service for the actual calculations.
 * @param {Object} context App context
 * @param {Object} order Relevant information about an order. This is similar to an OrderFulfillmentGroup type.
 * @param {Boolean} forceZeroes Set to `true` to force tax properties to be added
 *   and set to 0 when no tax plugin is enabled. When calculating tax for a cart, this should be false.
 *   When calculating tax for an order, this should be true.
 * @returns {Object} Calculated tax information. Has `taxSummary` property in `TaxSummary` schema
 *   as well as `itemTaxes` array property with `itemId`, `tax`, `taxableAmount`,
 *   and `taxes` properties on each array item.
 */
export default async function getFulfillmentGroupTaxes(context, { order, forceZeroes }) {
  try {
    CommonOrder.validate(order);
  } catch (error) {
    Logger.error("Invalid order input provided to getFulfillmentGroupTaxes", error);
    throw new ReactionError("internal-error", "Error while calculating taxes");
  }

  const { items, shopId } = order;

  const activeTaxService = await getActiveTaxServiceForShop(context, shopId);

  const defaultReturnValue = {
    taxSummary: {
      calculatedAt: new Date(),
      tax: 0,
      taxableAmount: 0,
      taxes: []
    },
    itemTaxes: items.map((item) => ({ itemId: item._id, tax: 0, taxableAmount: 0, taxes: [] }))
  };

  if (!activeTaxService) {
    return forceZeroes ? defaultReturnValue : { itemTaxes: [], taxSummary: null };
  }

  let taxServiceResult;
  try {
    taxServiceResult = await activeTaxService.functions.calculateOrderTaxes({ context, order });
  } catch (error) {
    Logger.error(`Error in calculateOrderTaxes for the active tax service (${activeTaxService.displayName})`, error);
    throw new ReactionError("internal-error", "Error while calculating taxes");
  }

  // The tax service may return `null` if it can't calculate due to missing info
  if (!taxServiceResult) {
    return forceZeroes ? defaultReturnValue : { itemTaxes: [], taxSummary: null };
  }

  try {
    TaxServiceResult.validate(taxServiceResult);
  } catch (error) {
    Logger.error(`Invalid return from calculateOrderTaxes for the active tax service (${activeTaxService.displayName})`, error);
    throw new ReactionError("internal-error", "Error while calculating taxes");
  }

  return taxServiceResult;
}
