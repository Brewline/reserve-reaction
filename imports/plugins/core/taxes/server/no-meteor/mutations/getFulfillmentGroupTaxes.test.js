import Factory from "/imports/test-utils/helpers/factory";
import mockContext from "/imports/test-utils/helpers/mockContext";
import getFulfillmentGroupTaxes from "./getFulfillmentGroupTaxes";
import { rewire$getTaxServicesForShop, restore as restoreRegistration } from "../registration";

const orderItem = Factory.CommonOrderItem.makeOne();

const order = Factory.CommonOrder.makeOne({
  items: [orderItem]
});
delete order._id; // There is a bug where Factory package adds _id even if it isn't in the schema

const getTaxServicesForShopMock = jest.fn().mockName("getTaxServicesForShop");

beforeAll(() => {
  rewire$getTaxServicesForShop(getTaxServicesForShopMock);
});

afterAll(() => {
  restoreRegistration();
});

test("no registered tax services, do not force zeroes", async () => {
  getTaxServicesForShopMock.mockReturnValueOnce(Promise.resolve({}));

  const result = await getFulfillmentGroupTaxes(mockContext, {
    forceZeroes: false,
    order
  });

  expect(result).toEqual({ itemTaxes: [], taxSummary: null });
});

test("no registered tax services, force zeroes", async () => {
  getTaxServicesForShopMock.mockReturnValueOnce(Promise.resolve({}));

  const result = await getFulfillmentGroupTaxes(mockContext, {
    forceZeroes: true,
    order
  });

  expect(result).toEqual({
    taxSummary: {
      calculatedAt: jasmine.any(Date),
      tax: 0,
      taxableAmount: 0,
      taxes: []
    },
    itemTaxes: [
      { itemId: orderItem._id, tax: 0, taxableAmount: 0, taxes: [] }
    ]
  });
});

test("null result from primary tax service, no fallback, do not force zeroes", async () => {
  const calculateOrderTaxes = jest.fn().mockName("calculateOrderTaxes").mockReturnValueOnce(Promise.resolve(null));

  getTaxServicesForShopMock.mockReturnValueOnce(Promise.resolve({
    primaryTaxService: {
      displayName: "Primary Tax Service",
      name: "primary-tax-service-mock",
      functions: {
        calculateOrderTaxes
      }
    }
  }));

  const result = await getFulfillmentGroupTaxes(mockContext, {
    forceZeroes: false,
    order
  });

  expect(result).toEqual({ itemTaxes: [], taxSummary: null });
});

test("null result from primary tax service, no fallback, force zeroes", async () => {
  const calculateOrderTaxes = jest.fn().mockName("calculateOrderTaxes").mockReturnValueOnce(Promise.resolve(null));

  getTaxServicesForShopMock.mockReturnValueOnce(Promise.resolve({
    primaryTaxService: {
      displayName: "Primary Tax Service",
      name: "primary-tax-service-mock",
      functions: {
        calculateOrderTaxes
      }
    }
  }));

  const result = await getFulfillmentGroupTaxes(mockContext, {
    forceZeroes: true,
    order
  });

  expect(result).toEqual({
    taxSummary: {
      calculatedAt: jasmine.any(Date),
      tax: 0,
      taxableAmount: 0,
      taxes: []
    },
    itemTaxes: [
      { itemId: orderItem._id, tax: 0, taxableAmount: 0, taxes: [] }
    ]
  });
});

test("null result from both tax services, do not force zeroes", async () => {
  const calculateOrderTaxes = jest.fn().mockName("calculateOrderTaxes").mockReturnValueOnce(Promise.resolve(null));
  const fallbackCalculateOrderTaxes = jest.fn().mockName("fallbackCalculateOrderTaxes").mockReturnValueOnce(Promise.resolve(null));

  getTaxServicesForShopMock.mockReturnValueOnce(Promise.resolve({
    fallbackTaxService: {
      displayName: "Fallback Tax Service",
      name: "fallback-tax-service-mock",
      functions: {
        calculateOrderTaxes: fallbackCalculateOrderTaxes
      }
    },
    primaryTaxService: {
      displayName: "Primary Tax Service",
      name: "primary-tax-service-mock",
      functions: {
        calculateOrderTaxes
      }
    }
  }));

  const result = await getFulfillmentGroupTaxes(mockContext, {
    forceZeroes: false,
    order
  });

  expect(result).toEqual({ itemTaxes: [], taxSummary: null });
});

test("null result from both tax services, force zeroes", async () => {
  const calculateOrderTaxes = jest.fn().mockName("calculateOrderTaxes").mockReturnValueOnce(Promise.resolve(null));
  const fallbackCalculateOrderTaxes = jest.fn().mockName("fallbackCalculateOrderTaxes").mockReturnValueOnce(Promise.resolve(null));

  getTaxServicesForShopMock.mockReturnValueOnce(Promise.resolve({
    fallbackTaxService: {
      displayName: "Fallback Tax Service",
      name: "fallback-tax-service-mock",
      functions: {
        calculateOrderTaxes: fallbackCalculateOrderTaxes
      }
    },
    primaryTaxService: {
      displayName: "Primary Tax Service",
      name: "primary-tax-service-mock",
      functions: {
        calculateOrderTaxes
      }
    }
  }));

  const result = await getFulfillmentGroupTaxes(mockContext, {
    forceZeroes: true,
    order
  });

  expect(result).toEqual({
    taxSummary: {
      calculatedAt: jasmine.any(Date),
      tax: 0,
      taxableAmount: 0,
      taxes: []
    },
    itemTaxes: [
      { itemId: orderItem._id, tax: 0, taxableAmount: 0, taxes: [] }
    ]
  });
});

test("minimal result from primary tax service", async () => {
  const taxes = [
    {
      _id: "mockTax",
      customFields: {
        taxesField: "abc123"
      },
      sourcing: "destination",
      tax: 15,
      taxName: "mockTaxName",
      taxRate: 0.05,
      taxableAmount: 90
    }
  ];

  const itemTax = {
    customFields: {
      taxesField: "abc123"
    },
    itemId: orderItem._id,
    tax: 15,
    taxableAmount: 90,
    taxes
  };

  const taxSummary = {
    calculatedAt: new Date(),
    calculatedByTaxServiceName: "primary-tax-service-mock",
    customFields: {
      taxesField: "abc123"
    },
    tax: 15,
    taxableAmount: 90,
    taxes
  };

  const mockResult = Factory.TaxServiceResult.makeOne({
    itemTaxes: [itemTax],
    taxSummary
  });
  delete mockResult._id; // There is a bug where Factory package adds _id even if it isn't in the schema

  const calculateOrderTaxes = jest.fn().mockName("calculateOrderTaxes").mockReturnValueOnce(Promise.resolve(mockResult));

  getTaxServicesForShopMock.mockReturnValueOnce(Promise.resolve({
    primaryTaxService: {
      displayName: "Primary Tax Service",
      name: "primary-tax-service-mock",
      functions: {
        calculateOrderTaxes
      }
    }
  }));

  const result = await getFulfillmentGroupTaxes(mockContext, {
    forceZeroes: false,
    order
  });

  expect(result).toEqual({
    itemTaxes: [itemTax],
    taxSummary
  });
});

test("null result from primary tax service, minimal result from fallback tax service", async () => {
  const taxes = [
    {
      _id: "mockTax",
      sourcing: "destination",
      tax: 15,
      taxName: "mockTaxName",
      taxRate: 0.05,
      taxableAmount: 90
    }
  ];

  const itemTax = {
    itemId: orderItem._id,
    tax: 15,
    taxableAmount: 90,
    taxes
  };

  const taxSummary = {
    calculatedAt: new Date(),
    calculatedByTaxServiceName: "primary-tax-service-mock",
    tax: 15,
    taxableAmount: 90,
    taxes
  };

  const mockResult = Factory.TaxServiceResult.makeOne({
    itemTaxes: [itemTax],
    taxSummary
  });
  delete mockResult._id; // There is a bug where Factory package adds _id even if it isn't in the schema

  const calculateOrderTaxes = jest.fn().mockName("calculateOrderTaxes").mockReturnValueOnce(Promise.resolve(null));
  const fallbackCalculateOrderTaxes = jest.fn().mockName("fallbackCalculateOrderTaxes").mockReturnValueOnce(Promise.resolve(mockResult));

  getTaxServicesForShopMock.mockReturnValueOnce(Promise.resolve({
    fallbackTaxService: {
      displayName: "Fallback Tax Service",
      name: "fallback-tax-service-mock",
      functions: {
        calculateOrderTaxes: fallbackCalculateOrderTaxes
      }
    },
    primaryTaxService: {
      displayName: "Primary Tax Service",
      name: "primary-tax-service-mock",
      functions: {
        calculateOrderTaxes
      }
    }
  }));

  const result = await getFulfillmentGroupTaxes(mockContext, {
    forceZeroes: false,
    order
  });

  expect(result).toEqual({
    itemTaxes: [itemTax],
    taxSummary
  });
});

test("throws if the result doesn't match the schema", async () => {
  const taxes = [
    {
      _id: "mockTax",
      tax: 15,
      taxName: "mockTaxName",
      taxRate: 0.05,
      taxableAmount: 90
    }
  ];

  const itemTax = {
    itemId: orderItem._id,
    tax: 15,
    taxableAmount: 90,
    taxes
  };

  const taxSummary = {
    calculatedAt: new Date(),
    calculatedByTaxServiceName: "primary-tax-service-mock",
    tax: 15,
    taxableAmount: 90,
    taxes
  };

  const mockResult = Factory.TaxServiceResult.makeOne({
    itemTaxes: [itemTax],
    taxSummary
  });
  delete mockResult._id; // There is a bug where Factory package adds _id even if it isn't in the schema

  const calculateOrderTaxes = jest.fn().mockName("calculateOrderTaxes").mockReturnValueOnce(Promise.resolve(mockResult));

  getTaxServicesForShopMock.mockReturnValueOnce(Promise.resolve({
    primaryTaxService: {
      displayName: "Primary Tax Service",
      name: "primary-tax-service-mock",
      functions: {
        calculateOrderTaxes
      }
    }
  }));

  const result = getFulfillmentGroupTaxes(mockContext, { forceZeroes: false, order });
  expect(result).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the calculation function throws", async () => {
  const calculateOrderTaxes = jest.fn().mockName("calculateOrderTaxes").mockImplementationOnce(async () => {
    throw new Error("Internal problem");
  });

  getTaxServicesForShopMock.mockReturnValueOnce(Promise.resolve({
    primaryTaxService: {
      displayName: "Primary Tax Service",
      name: "primary-tax-service-mock",
      functions: {
        calculateOrderTaxes
      }
    }
  }));

  const result = getFulfillmentGroupTaxes(mockContext, { forceZeroes: false, order });
  expect(result).rejects.toThrowErrorMatchingSnapshot();
});
