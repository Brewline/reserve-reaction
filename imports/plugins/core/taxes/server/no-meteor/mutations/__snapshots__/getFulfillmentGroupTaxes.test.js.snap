// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if the calculation function throws 1`] = `"Error while calculating taxes"`;

exports[`throws if the result doesn't match the schema 1`] = `"Error while calculating taxes"`;
